README.md
[![CI status](https://code.ffdn.org/zorun/kea-hook-runscript/badges/master/pipeline.svg)](https://code.ffdn.org/zorun/kea-hook-runscript/-/pipelines)

# About kea-hook-runscript

This a hook for the Kea DHCP server that allows to run an external script
at various points in the processing of DHCP requests and responses.

The goal is to simplify integration with Kea: for many simple use-cases,
it is overkill to have to write a full-blown Kea hook, where a simple shell script
can do the job.

This hook is licensed under the Mozilla Public License version 2 (MPL2).

## What it can and can't do

Integration is mostly done one-way: thanks to this hook, Kea passes information
to the external script, but the script cannot easily modify Kea's behaviour.

The **external script** can be any kind of executable program, but often it will
be a simple script (shell, Perl, Python...).  Information about what Kea is doing
is provided to the external script through environment variables: MAC address of
the requesting DHCP client, IP address being handed out, etc.

Each time Kea encounters a hook point, it will (by default) call the script
**synchronously**.
That is, Kea will do absolutely nothing else while the script is running.
Thus, it is a good idea to perform only lightweight processing in the script,
and absolutely avoid blocking operations.  Also, scripting languages that need
to initialise a huge interpreter (such as Python or Ruby) will cause a large
amount of CPU usage and a massive slowdown of Kea, because the script is run
multiple times for each DHCP transaction.

If you know what you are doing, you can optionally call the script **asynchronously**
by setting `wait` to `false` (see below).

This hook works for both DHCPv4 and DHCPv6, on Kea 1.1 and above.

In the future, the hook will possibly feed the return code of the external script
back into Kea.  This would allow the external script to cancel part of Kea's normal processing
(for instance, it could be possible to easily implement a flexible host blacklist this way).

## Use-cases

Given the limitations exposed above, here are some example use-cases for which
this hook is well-suited:

- add/remove routing entries when DHCP clients arrive or leave.  This can be useful
  when handing out IPv4 addressing in /32 subnets, or IPv6 Prefix Delegation
  with DHCPv6-PD.  An example is included in `examples/slash32_leases/`;
- update firewall rules to allow/refuse access to new DHCP clients;
- log information about successful leases.

For more complex use-cases, including non-trivial changes to Kea's behaviour,
it may be easier to just write a Kea hook yourself.

## Examples

If you have more examples of usage, feel free to contribute your Kea
config and your scripts!

### Managing routes for IPv6 delegated prefixes

When delegating IPv6 prefixes with DHCPv6-PD, it is necessary to add the corresponding routes
in the kernel.

This example script adds/removes static IPv6 routes whenever Kea delegates an IPv6 prefix
through DHCPv6-PD or when the lease expires.

See the included [README](examples/ipv6_prefix_delegation/README.md) for more
explanations and the [source](examples/ipv6_prefix_delegation) with the script and
an example Kea configuration.

### Handing out IPv4 addresses in /32 subnets

This example allows to lease IPv4 addresses individually (/32 subnets), by
inserting routes in the kernel each time a DHCP client connects, and
sending custom routes to clients using DHCP option 121.  This is mostly
useful to hand out public IPv4 addresses to customers.

See the included [README](examples/slash32_leases/README.md) for more
explanations and the [source](examples/slash32_leases) with the script and
an example Kea configuration.

### Debug script

To experiment, a simple debug script is provided: `examples/debug.sh`.  It
simply prints the name of the hook point and all environment variables
passed to it.

The output of the script is at `/tmp/kea-hook-runscript-debug.log`.  A nice way to debug
is to continously display the content of this file:

    tail -F /tmp/kea-hook-runscript-debug.log

## Pre-built binaries

Since version 1.4.0, we have a CI system to build the hook on various OS and for various
versions of Kea. It's new, so there might be bugs.

The binaries are available from the [release page](https://github.com/zorun/kea-hook-runscript/releases)
or you can directly
[browse through the pipeline results](https://code.ffdn.org/zorun/kea-hook-runscript/-/pipelines).

## How to build

If you want to build the hook yourself, you need the Kea libraries as well
as the Kea and Boost development headers.

### Using a packaged version of Kea

If you use a Kea package, you need the appropriate development packages:

- boost development files: `libboost-dev` or equivalent
- kea development files: `isc-kea-dev` from cloudsmith (official Kea package)

If you prefer using the Kea package from Debian, install `kea-dev` instead.
However, it is currently unsupported and is only available in sid.

Then, to build the hook, simply run:

    $ make -j4

### Using Kea source

To build against a local Kea source tree, assumed to be in `~/kea`:

- build Kea (`cd ~/kea && make -j`)
- install Kea to a local directory (`cd ~/kea && make install DESTDIR=/tmp/kea`)

Then build this hook with:

    $ export KEA_INCLUDE=$HOME/kea/src/lib
    $ export KEA_LIB=/tmp/kea/usr/local/lib
    $ make

### Supported Kea versions

Some notes on Kea versions:

- Kea 1.1 does not install all required headers (most notably `dhcpsrv/`),
  so you may need to build against Kea's source tree.
- Kea 1.2 is missing a header file by mistake, so depending on your
  distribution, you may need to manually copy `option6_pdexclude.h` from
  the Kea git repository to `/usr/include/kea/dhcp/`.
- Kea 1.3 to 1.7 should work out-of-the-box.
- Kea 1.8 needs to run without [multi-threading](https://kea.readthedocs.io/en/kea-1.8.0/arm/dhcp4-srv.html#multi-threading-settings).
  Open a ticket if you need multi-threading support.

## How to use this hook

If all goes well, you should obtain a `kea-hook-runscript.so` file.
Then, here is how to tell Kea to use this hook, for DHCPv4:

    {
    "Dhcp4":
    {
      "hooks-libraries": [
        {
          "library": "/path/to/hea-hook-runscript/kea-hook-runscript.so",
          "parameters": {
            "script": "/path/to/myscript.sh",
            "wait": true
          }
        }
      ],
      ...
    }
    }

The `wait` parameter indicates whether Kea waits for the script to exit.  That is,
if set to `true`, Kea will block while the script is running.
If you need high-performance DHCP, you can set it to `false`, but you must be prepared
to handle several instances of the script running in parallel.

You can use the same script for both DHCPv4 and DHCPv6, or use two different scripts.

The script will receive the name of the hook point as first argument, and all
relevant information available at the current hook point will be passed as
environment variables, documented below.

To debug, see the `examples/debug.sh` script described above.

Refer to the Kea documentation for more information about each hook point:

- DHCPv4 hooks reference: <https://jenkins.isc.org/job/Kea_doc/doxygen/de/df3/dhcpv4Hooks.html>
- DHCPv6 hooks reference: <https://jenkins.isc.org/job/Kea_doc/doxygen/d1/d02/dhcpv6Hooks.html>

# Reference of variables passed to the external script

## DHCPv4 variables

Here are all possible variables for DHCPv4, with their type, description
and reference of the possible values.  Booleans are simply expressed with
`0` and `1`.

| Variable name                         | Type     | Description                                                 | Reference                                                                                                                                                       |
|---------------------------------------|----------|-------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `KEA_QUERY4_TYPE`                     | `string` | Type of DHCP message                                        | [dhcp/dhcp4.h](https://jenkins.isc.org/job/Kea_doc/doxygen/d5/d8c/namespaceisc_1_1dhcp.html#aa5bfdcc4861aa3dab5328dba89362016)                                  |
| `KEA_QUERY4_INTERFACE`                | `string` | Interface on which query was received                       |                                                                                                                                                                 |
| `KEA_QUERY4_IFINDEX`                  | `int`    | Index of the interface on which query was received          |                                                                                                                                                                 |
| `KEA_QUERY4_HWADDR`                   | `string` | Hardware address of the client (its MAC address)            |                                                                                                                                                                 |
| `KEA_QUERY4_HWADDR_TYPE`              | `int`    | Type of hardware address                                    | [dhcp/dhcp4.h](https://jenkins.isc.org/job/Kea_doc/doxygen/d5/d8c/namespaceisc_1_1dhcp.html#addcff933049489d800f9869196c8e46fa96a62c59182d6e06780b0e1ef40da059) |
| `KEA_QUERY4_HWADDR_SOURCE`            | `int`    | How this MAC address was obtained                           | [dhcp/hwaddr.h](https://jenkins.isc.org/job/Kea_doc/doxygen/da/dae/group__hw__sources.html)                                                                     |
| `KEA_QUERY4_RELAYED`                  | `bool`   | Whether query was relayed                                   | [dhcp/pkt4.h](https://jenkins.isc.org/job/Kea_doc/doxygen/de/d13/classisc_1_1dhcp_1_1Pkt4.html#a8468401827b9bacdd3796bb4e20d8e5e)                               |
| `KEA_QUERY4_RELAY_HOPS`               | `int`    | Number of relay agents traversed                            |                                                                                                                                                                 |
| `KEA_QUERY4_OPTION60`                 | `string` | Option 60 - vendor id                                       |                                                                                                                                                                 |
| `KEA_QUERY4_CIADDR`                   | `string` | Client IP address                                           | [dhcp/pkt4.h](https://jenkins.isc.org/job/Kea_doc/doxygen/de/d13/classisc_1_1dhcp_1_1Pkt4.html)                                                                 |
| `KEA_QUERY4_SIADDR`                   | `string` | Server IP address                                           | [dhcp/pkt4.h](https://jenkins.isc.org/job/Kea_doc/doxygen/de/d13/classisc_1_1dhcp_1_1Pkt4.html)                                                                 |
| `KEA_QUERY4_YIADDR`                   | `string` | Your IP address                                             | [dhcp/pkt4.h](https://jenkins.isc.org/job/Kea_doc/doxygen/de/d13/classisc_1_1dhcp_1_1Pkt4.html)                                                                 |
| `KEA_QUERY4_GIADDR`                   | `string` | Gateway IP address (inserted by DHCP relay)                 | [dhcp/pkt4.h](https://jenkins.isc.org/job/Kea_doc/doxygen/de/d13/classisc_1_1dhcp_1_1Pkt4.html)                                                                 |
| `KEA_QUERY4_RAI`                      | `string` | Relay Agent Information (RFC 3046) as hex string            |                                                                                                                                                                 |
| `KEA_QUERY4_RAI_CIRCUIT_ID`           | `string` | RAI sub-option 1 Circuit id (RFC 3046) as hex string        |                                                                                                                                                                 |
| `KEA_QUERY4_RAI_REMOTE_ID`            | `string` | RAI sub-option 2 Remote id (RFC 3046) as hex string         |                                                                                                                                                                 |
| `KEA_RESPONSE4_TYPE`                  | `string` | Type of DHCP message                                        | [dhcp/dhcp4.h](https://jenkins.isc.org/job/Kea_doc/doxygen/d5/d8c/namespaceisc_1_1dhcp.html#aa5bfdcc4861aa3dab5328dba89362016)                                  |
| `KEA_RESPONSE4_INTERFACE`             | `string` | Interface on which response is being sent                   |                                                                                                                                                                 |
| `KEA_RESPONSE4_IFINDEX`               | `int`    | Index of the interface on which response is being sent      |                                                                                                                                                                 |
| `KEA_RESPONSE4_HWADDR`                | `string` | Hardware address of the client (its MAC address)            |                                                                                                                                                                 |
| `KEA_RESPONSE4_HWADDR_TYPE`           | `int`    | Type of hardware address                                    | [dhcp/dhcp4.h](https://jenkins.isc.org/job/Kea_doc/doxygen/d5/d8c/namespaceisc_1_1dhcp.html#addcff933049489d800f9869196c8e46fa96a62c59182d6e06780b0e1ef40da059) |
| `KEA_RESPONSE4_HWADDR_SOURCE`         | `int`    | How this MAC address was obtained                           | [dhcp/hwaddr.h](https://jenkins.isc.org/job/Kea_doc/doxygen/da/dae/group__hw__sources.html)                                                                     |
| `KEA_RESPONSE4_RELAYED`               | `bool`   | Whether response is relayed                                 | [dhcp/pkt4.h](https://jenkins.isc.org/job/Kea_doc/doxygen/de/d13/classisc_1_1dhcp_1_1Pkt4.html#a8468401827b9bacdd3796bb4e20d8e5e)                               |
| `KEA_RESPONSE4_RELAY_HOPS`            | `int`    | Number of relay agents traversed                            |                                                                                                                                                                 |
| `KEA_RESPONSE4_CIADDR`                | `string` | Client IP address                                           | [dhcp/pkt4.h](https://jenkins.isc.org/job/Kea_doc/doxygen/de/d13/classisc_1_1dhcp_1_1Pkt4.html)                                                                 |
| `KEA_RESPONSE4_SIADDR`                | `string` | Server IP address                                           | [dhcp/pkt4.h](https://jenkins.isc.org/job/Kea_doc/doxygen/de/d13/classisc_1_1dhcp_1_1Pkt4.html)                                                                 |
| `KEA_RESPONSE4_YIADDR`                | `string` | Your IP address                                             | [dhcp/pkt4.h](https://jenkins.isc.org/job/Kea_doc/doxygen/de/d13/classisc_1_1dhcp_1_1Pkt4.html)                                                                 |
| `KEA_RESPONSE4_GIADDR`                | `string` | Gateway IP address                                          | [dhcp/pkt4.h](https://jenkins.isc.org/job/Kea_doc/doxygen/de/d13/classisc_1_1dhcp_1_1Pkt4.html)                                                                 |
| `KEA_SUBNET4_PREFIX`                  | `IPv4`   | IP prefix of the subnet (without prefix length)             |                                                                                                                                                                 |
| `KEA_SUBNET4_PREFIXLEN`               | `int`    | Prefix length of the subnet (`0` to `32`)                   |                                                                                                                                                                 |
| `KEA_SUBNET4`                         | `string` | `KEA_SUBNET4_PREFIX`/`KEA_SUBNET4_PREFIXLEN`                |                                                                                                                                                                 |
| `KEA_LEASE4_ADDRESS`                  | `IPv4`   | IPv4 address leased to client                               |                                                                                                                                                                 |
| `KEA_LEASE4_TYPE`                     | `string` | Type of lease, always equal to "V4"                         |                                                                                                                                                                 |
| `KEA_LEASE4_HWADDR`                   | `string` | Hardware address of the client                              |                                                                                                                                                                 |
| `KEA_LEASE4_HOSTNAME`                 | `string` | Hostname associated to the client                           |                                                                                                                                                                 |

CHANGELOG.md
# Changelog for kea-hook-runscript

## 1.4.0 (2021-01-08)

New features:

- Add support for DHCP option 82, Relay Agent Information (#22)
- Generate binaries with a [CI system](https://code.ffdn.org/zorun/kea-hook-runscript/-/pipelines)
  Currently, Debian and Ubuntu are supported, for Kea versions from 1.6 to 1.9.
  Binaries are attached to this release for convenience.

Fixes:

- Ship pre-built `messages.{cc,h}` to avoid build dependency on kea-msg-compiler.
  This should make it easier to build the hook from git.

Misc:

- Add IPv6 prefix delegation example

## 1.3.3 (2020-02-27)

Fixes:

- Fix a crash in `lease6_expire` when IPv6 leases are reclaimed by Kea (#18)

## 1.3.2 (2020-01-31)

Fixes:

- Fix MAC address reported in `KEA_QUERY4_HWADDR` when the query is received from a DHCP relay (#17)

## 1.3.1 (2019-11-02)

Misc:

- Produce release tarballs with pre-built messages, to avoid depending on the `kea-msg-compiler` binary (#15)

## 1.3.0 (2019-11-02)

New features:

- Add support for DHCP option 60 (#12)

Fixes:

- Fix build against Kea 1.6 (#16)
- Fix build with older compilers

Misc:

- Licence code under the MPL

## 1.2.0 (2019-01-24)

- Allow to run the script asynchronously (by setting `wait=false`, see `README`)

## 1.1.0 (2017-10-22)

- Add lots of DHCPv6 variables: DHCPv6 should now be usable
- Add some more DHCPv4 variables, especially for leases
- Safety fixes: fix potential NULL pointer dereference, and unsafe string usage
- Makefile: use standard CXX variable instead of hard-coding g++

## 1.0.0 (2017-08-24)

- Initial release


src/callouts.cc
/* Copyright (c) 2017-2019 by Baptiste Jonglez
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

#include <hooks/hooks.h>
#include <dhcp/pkt4.h>
#include <dhcp/dhcp6.h>
#include <dhcp/pkt6.h>
#include <dhcp/option6_ia.h>
#include <dhcpsrv/subnet.h>
#include <dhcpsrv/lease.h>

#include <string>
#include <vector>

#include "runscript.h"

using namespace isc::dhcp;
using namespace isc::hooks;

extern "C" {

/* These are helpers that extract relevant information from Kea data
 * structures and store them in environment variables. */
void extract_bool(std::vector<std::string>& env, const std::string variable, bool value)
{
    env.push_back(variable + "=" + std::string(value ? "1" : "0"));
}

/* Extract information from a DHCPv4 packet (query received, or response
 * about to be sent) */
void extract_pkt4(std::vector<std::string>& env, const std::string envprefix, const Pkt4Ptr pkt4)
{
    /* General information */
    env.push_back(envprefix + "TYPE=" + std::string(pkt4->getName()));
    env.push_back(envprefix + "INTERFACE=" + pkt4->getIface());
    env.push_back(envprefix + "IFINDEX=" + std::to_string(pkt4->getIndex()));
    /* Hardware address */
    HWAddrPtr hwaddr = pkt4->getHWAddr();
    if (hwaddr) {
        env.push_back(envprefix + "HWADDR=" + hwaddr->toText(false));
        env.push_back(envprefix + "HWADDR_TYPE=" + std::to_string(hwaddr->htype_));
        env.push_back(envprefix + "HWADDR_SOURCE=" + std::to_string(hwaddr->source_));
    } else {
        env.push_back(envprefix + "HWADDR=");
        env.push_back(envprefix + "HWADDR_TYPE=");
        env.push_back(envprefix + "HWADDR_SOURCE=");
    }
    /* Misc */
    env.push_back(envprefix + "CIADDR=" + pkt4->getCiaddr().toText());
    env.push_back(envprefix + "SIADDR=" + pkt4->getSiaddr().toText());
    env.push_back(envprefix + "YIADDR=" + pkt4->getYiaddr().toText());
    env.push_back(envprefix + "GIADDR=" + pkt4->getGiaddr().toText());
    env.push_back(envprefix + "RELAYED=" + std::to_string(pkt4->isRelayed()));
    env.push_back(envprefix + "RELAY_HOPS=" + std::to_string(pkt4->getHops()));

    /* Specific Options */
    OptionPtr option60 = pkt4->getOption(60);
    if (option60) {
        env.push_back(envprefix + "OPTION60=" + option60->toString());
    }

    OptionPtr rai = pkt4->getOption(82);
    if (rai) {
        env.push_back(envprefix + "RAI=" + rai->toHexString());

        OptionPtr circuit_id = rai->getOption(RAI_OPTION_AGENT_CIRCUIT_ID);
        if (circuit_id) {
            env.push_back(envprefix + "RAI_CIRCUIT_ID=" + circuit_id->toHexString());
        }

        OptionPtr remote_id = rai->getOption(RAI_OPTION_REMOTE_ID);
        if (remote_id) {
            env.push_back(envprefix + "RAI_REMOTE_ID=" + remote_id->toHexString());
        }
    }
}

void extract_query4(std::vector<std::string>& env, const Pkt4Ptr query)
{
    extract_pkt4(env, "KEA_QUERY4_", query);
}

void extract_response4(std::vector<std::string>& env, const Pkt4Ptr response)
{
    extract_pkt4(env, "KEA_RESPONSE4_", response);
}

/* Extract information from a DHCPv6 packet (query received, or response
 * about to be sent) */
void extract_pkt6(std::vector<std::string>& env, const std::string envprefix, const Pkt6Ptr pkt6)
{
    /* General information */
    env.push_back(envprefix + "TYPE=" + std::string(pkt6->getName()));
    env.push_back(envprefix + "INTERFACE=" + pkt6->getIface());
    env.push_back(envprefix + "IFINDEX=" + std::to_string(pkt6->getIndex()));
    HWAddrPtr hwaddr = pkt6->getMAC(HWAddr::HWADDR_SOURCE_ANY);
    if (hwaddr) {
        env.push_back(envprefix + "HWADDR=" + hwaddr->toText(false));
        env.push_back(envprefix + "HWADDR_TYPE=" + std::to_string(hwaddr->htype_));
        env.push_back(envprefix + "HWADDR_SOURCE=" + std::to_string(hwaddr->source_));
    } else {
        env.push_back(envprefix + "HWADDR=");
        env.push_back(envprefix + "HWADDR_TYPE=");
        env.push_back(envprefix + "HWADDR_SOURCE=");
    }
    env.push_back(envprefix + "LOCAL_ADDRESS=" + pkt6->getLocalAddr().toText());
    env.push_back(envprefix + "LOCAL_PORT=" + std::to_string(pkt6->getLocalPort()));
    env.push_back(envprefix + "REMOTE_ADDRESS=" + pkt6->getRemoteAddr().toText());
    env.push_back(envprefix + "REMOTE_PORT=" + std::to_string(pkt6->getRemotePort()));
    env.push_back(envprefix + "LABEL=" + pkt6->getLabel());
    env.push_back(envprefix + "TRANSACTION_ID=" + std::to_string(pkt6->getTransid()));
    /* TODO */
    env.push_back(envprefix + "DUID=");
    /* TODO: all options?  Only common ones?  Which format? */
    /* TODO */
    env.push_back(envprefix + "DEBUG=" + pkt6->toText());
}

void extract_query6(std::vector<std::string>& env, const Pkt6Ptr query)
{
    extract_pkt6(env, "KEA_QUERY6_", query);
}


void extract_response6(std::vector<std::string>& env, const Pkt6Ptr response)
{
    extract_pkt6(env, "KEA_RESPONSE6_", response);
}

void extract_subnet4(std::vector<std::string>& env, const Subnet4Ptr subnet)
{
    /* The subnet given by Kea might be NULL, this seems to happen when
     * Kea fails to find a matching subnet for a client request. */
    if (subnet != NULL) {
        env.push_back("KEA_SUBNET4=" + subnet->toText());
        std::pair<isc::asiolink::IOAddress, uint8_t> prefix = subnet->get();
        env.push_back("KEA_SUBNET4_PREFIX=" + prefix.first.toText());
        env.push_back("KEA_SUBNET4_PREFIXLEN=" + std::to_string(prefix.second));
    } else {
        env.push_back("KEA_SUBNET4=");
        env.push_back("KEA_SUBNET4_PREFIX=");
        env.push_back("KEA_SUBNET4_PREFIXLEN=");
    }
}

void extract_subnet6(std::vector<std::string>& env, const Subnet6Ptr subnet)
{
    if (subnet != NULL) {
        env.push_back("KEA_SUBNET6=" + subnet->toText());
        std::pair<isc::asiolink::IOAddress, uint8_t> prefix = subnet->get();
        env.push_back("KEA_SUBNET6_PREFIX=" + prefix.first.toText());
        env.push_back("KEA_SUBNET6_PREFIXLEN=" + std::to_string(prefix.second));
    } else {
        env.push_back("KEA_SUBNET6=");
        env.push_back("KEA_SUBNET6_PREFIX=");
        env.push_back("KEA_SUBNET6_PREFIXLEN=");
    }
}

void extract_lease4(std::vector<std::string>& env, const Lease4Ptr lease)
{
    env.push_back("KEA_LEASE4_TYPE=V4");
    env.push_back("KEA_LEASE4_STATE=" + lease->basicStatesToText(lease->state_));
    extract_bool(env, "KEA_LEASE4_IS_EXPIRED", lease->expired());
    env.push_back("KEA_LEASE4_ADDRESS=" + lease->addr_.toText());
    if (lease->hwaddr_) {
        env.push_back("KEA_LEASE4_HWADDR=" + lease->hwaddr_->toText(false));
    } else {
        env.push_back("KEA_LEASE4_HWADDR=");
    }
    env.push_back("KEA_LEASE4_HOSTNAME=" + lease->hostname_);
    env.push_back("KEA_LEASE4_CLIENT_LAST_TRANSMISSION=" + std::to_string(lease->cltt_));
    env.push_back("KEA_LEASE4_VALID_LIFETIME=" + std::to_string(lease->valid_lft_));
    env.push_back("KEA_LEASE4_DEBUG=" + lease->toText());
}

void extract_lease6(std::vector<std::string>& env, const Lease6Ptr lease)
{
    env.push_back("KEA_LEASE6_TYPE=" + lease->typeToText(lease->type_));
    env.push_back("KEA_LEASE6_STATE=" + lease->basicStatesToText(lease->state_));
    extract_bool(env, "KEA_LEASE6_IS_EXPIRED", lease->expired());
    env.push_back("KEA_LEASE6_ADDRESS=" + lease->addr_.toText());
    if (lease->type_ == Lease::TYPE_PD) {
        env.push_back("KEA_LEASE6_DELEGATED_PREFIX=" + lease->addr_.toText() + "/" + std::to_string(lease->prefixlen_));
        env.push_back("KEA_LEASE6_DELEGATED_PREFIXLEN=" + std::to_string(lease->prefixlen_));
    }
    if (lease->hwaddr_) {
        env.push_back("KEA_LEASE6_HWADDR=" + lease->hwaddr_->toText(false));
    } else {
        env.push_back("KEA_LEASE6_HWADDR=");
    }
    env.push_back("KEA_LEASE6_HOSTNAME=" + lease->hostname_);
    env.push_back("KEA_LEASE6_CLIENT_DUID=" + lease->duid_->toText());
    env.push_back("KEA_LEASE6_CLIENT_LAST_TRANSMISSION=" + std::to_string(lease->cltt_));
    env.push_back("KEA_LEASE6_VALID_LIFETIME=" + std::to_string(lease->valid_lft_));
    env.push_back("KEA_LEASE6_PREFERRED_LIFETIME=" + std::to_string(lease->preferred_lft_));
    env.push_back("KEA_LEASE6_IAID=" + std::to_string(lease->iaid_));
    env.push_back("KEA_LEASE6_DEBUG=" + lease->toText());
}

/* IPv4 callouts */
int pkt4_receive(CalloutHandle& handle) {
    std::vector<std::string> env;
    Pkt4Ptr query;
    handle.getArgument("query4", query);
    extract_query4(env, query);
    /* Run script */
    int ret;
    ret = run_script("pkt4_receive", env);
    return 0;
}

int pkt4_send(CalloutHandle& handle) {
    std::vector<std::string> env;
    Pkt4Ptr response;
    Pkt4Ptr query;
    handle.getArgument("response4", response);
    extract_response4(env, response);
    handle.getArgument("query4", query);
    extract_query4(env, query);
    /* Run script */
    int ret;
    ret = run_script("pkt4_send", env);
    return 0;
}

int subnet4_select(CalloutHandle& handle) {
    std::vector<std::string> env;
    Pkt4Ptr query;
    Subnet4Ptr subnet;
    handle.getArgument("query4", query);
    extract_query4(env, query);
    handle.getArgument("subnet4", subnet);
    extract_subnet4(env, subnet);
    /* Run script */
    int ret;
    ret = run_script("subnet4_select", env);
    return 0;
}

int lease4_select(CalloutHandle& handle) {
    std::vector<std::string> env;
    Pkt4Ptr query;
    Subnet4Ptr subnet;
    bool fake_allocation;
    Lease4Ptr lease;
    handle.getArgument("query4", query);
    extract_query4(env, query);
    handle.getArgument("subnet4", subnet);
    extract_subnet4(env, subnet);
    handle.getArgument("fake_allocation", fake_allocation);
    extract_bool(env, "KEA_FAKE_ALLOCATION", fake_allocation);
    handle.getArgument("lease4", lease);
    extract_lease4(env, lease);
    /* Run script */
    int ret;
    ret = run_script("lease4_select", env);
    return 0;
}

int lease4_renew(CalloutHandle& handle) {
    std::vector<std::string> env;
    Pkt4Ptr query;
    Subnet4Ptr subnet;
    Lease4Ptr lease;
    handle.getArgument("query4", query);
    extract_query4(env, query);
    handle.getArgument("subnet4", subnet);
    extract_subnet4(env, subnet);
    /* TODO: what is this?  Is it different from what is in the query? */
    //handle.getArgument("clientid", XX);
    //handle.getArgument("hwaddr", XX);
    handle.getArgument("lease4", lease);
    extract_lease4(env, lease);
    /* Run script */
    int ret;
    ret = run_script("lease4_renew", env);
    return 0;
}

int lease4_release(CalloutHandle& handle) {
    std::vector<std::string> env;
    Pkt4Ptr query;
    Lease4Ptr lease;
    handle.getArgument("query4", query);
    extract_query4(env, query);
    handle.getArgument("lease4", lease);
    extract_lease4(env, lease);
    /* Run script */
    int ret;
    ret = run_script("lease4_release", env);
    return 0;
}

int lease4_decline(CalloutHandle& handle) {
    std::vector<std::string> env;
    Pkt4Ptr query;
    Lease4Ptr lease;
    handle.getArgument("query4", query);
    extract_query4(env, query);
    handle.getArgument("lease4", lease);
    extract_lease4(env, lease);
    /* Run script */
    int ret;
    ret = run_script("lease4_decline", env);
    return 0;
}

int lease4_expire(CalloutHandle& handle) {
    std::vector<std::string> env;
    Lease4Ptr lease;
    bool remove_lease;
    handle.getArgument("lease4", lease);
    extract_lease4(env, lease);
    handle.getArgument("remove_lease", remove_lease);
    extract_bool(env, "KEA_REMOVE_LEASE", remove_lease);
    /* Run script */
    int ret;
    ret = run_script("lease4_expire", env);
    return 0;
}

int lease4_recover(CalloutHandle& handle) {
    std::vector<std::string> env;
    Lease4Ptr lease;
    handle.getArgument("lease4", lease);
    extract_lease4(env, lease);
    /* Run script */
    int ret;
    ret = run_script("lease4_recover", env);
    return 0;
}

/* IPv6 callouts */
int pkt6_receive(CalloutHandle& handle) {
    std::vector<std::string> env;
    Pkt6Ptr query;
    handle.getArgument("query6", query);
    extract_query6(env, query);
    /* Run script */
    int ret;
    ret = run_script("pkt6_receive", env);
    return 0;
}

int pkt6_send(CalloutHandle& handle) {
    std::vector<std::string> env;
    Pkt6Ptr query, response;
    handle.getArgument("query6", query);
    extract_query6(env, query);
    handle.getArgument("response6", response);
    extract_response6(env, response);
    /* Run script */
    int ret;
    ret = run_script("pkt6_send", env);
    return 0;
}

int subnet6_select(CalloutHandle& handle) {
    std::vector<std::string> env;
    Pkt6Ptr query;
    Subnet6Ptr subnet;
    handle.getArgument("query6", query);
    extract_query6(env, query);
    handle.getArgument("subnet6", subnet);
    extract_subnet6(env, subnet);
    /* Run script */
    int ret;
    ret = run_script("subnet6_select", env);
    return 0;
}

int lease6_select(CalloutHandle& handle) {
    std::vector<std::string> env;
    Pkt6Ptr query;
    Subnet6Ptr subnet;
    bool fake_allocation;
    Lease6Ptr lease;
    handle.getArgument("query6", query);
    extract_query6(env, query);
    handle.getArgument("subnet6", subnet);
    extract_subnet6(env, subnet);
    handle.getArgument("fake_allocation", fake_allocation);
    extract_bool(env, "KEA_FAKE_ALLOCATION", fake_allocation);
    handle.getArgument("lease6", lease);
    extract_lease6(env, lease);
    /* Run script */
    int ret;
    ret = run_script("lease6_select", env);
    return 0;
}

int lease6_renew(CalloutHandle& handle) {
    std::vector<std::string> env;
    Pkt6Ptr query;
    Lease6Ptr lease;
    boost::shared_ptr<Option6IA> ia_na;
    boost::shared_ptr<Option6IA> ia_pd;
    handle.getArgument("query6", query);
    extract_query6(env, query);
    handle.getArgument("lease6", lease);
    extract_lease6(env, lease);
    try {
        handle.getArgument("ia_na", ia_na);
        /* TODO: use ia_na */
    } catch (const NoSuchArgument&) { }
    try {
        handle.getArgument("ia_pd", ia_pd);
        /* TODO: use ia_pd */
    } catch (const NoSuchArgument&) { }
    /* Run script */
    int ret;
    ret = run_script("lease6_renew", env);
    return 0;
}

int lease6_rebind(CalloutHandle& handle) {
    std::vector<std::string> env;
    Pkt6Ptr query;
    Lease6Ptr lease;
    boost::shared_ptr<Option6IA> ia_na;
    boost::shared_ptr<Option6IA> ia_pd;
    handle.getArgument("query6", query);
    extract_query6(env, query);
    handle.getArgument("lease6", lease);
    extract_lease6(env, lease);
    try {
        handle.getArgument("ia_na", ia_na);
        /* TODO: use ia_na */
    } catch (const NoSuchArgument&) { }
    try {
        handle.getArgument("ia_pd", ia_pd);
        /* TODO: use ia_pd */
    } catch (const NoSuchArgument&) { }
    /* Run script */
    int ret;
    ret = run_script("lease6_rebind", env);
    return 0;
}

int lease6_decline(CalloutHandle& handle) {
    std::vector<std::string> env;
    Pkt6Ptr query;
    Lease6Ptr lease;
    handle.getArgument("query6", query);
    extract_query6(env, query);
    handle.getArgument("lease6", lease);
    extract_lease6(env, lease);
    /* Run script */
    int ret;
    ret = run_script("lease6_decline", env);
    return 0;
}

int lease6_release(CalloutHandle& handle) {
    std::vector<std::string> env;
    Pkt6Ptr query;
    Lease6Ptr lease;
    handle.getArgument("query6", query);
    extract_query6(env, query);
    handle.getArgument("lease6", lease);
    extract_lease6(env, lease);
    /* Run script */
    int ret;
    ret = run_script("lease6_release", env);
    return 0;
}

int lease6_expire(CalloutHandle& handle) {
    std::vector<std::string> env;
    Lease6Ptr lease;
    bool remove_lease;
    handle.getArgument("lease6", lease);
    extract_lease6(env, lease);
    handle.getArgument("remove_lease", remove_lease);
    extract_bool(env, "KEA_REMOVE_LEASE", remove_lease);
    /* Run script */
    int ret;
    ret = run_script("lease6_expire", env);
    return 0;
}

int lease6_recover(CalloutHandle& handle) {
    std::vector<std::string> env;
    Lease6Ptr lease;
    handle.getArgument("lease6", lease);
    extract_lease6(env, lease);
    /* Run script */
    int ret;
    ret = run_script("lease6_recover", env);
    return 0;
}

} // end extern "C"
